utils/modeler/modeler.go

```diff
@@ -0,0 +1,107 @@
+package modeler
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+// Modeler is a utility for populating an arbitrary model's fields with values from a
+// map[string]string.
+type Modeler struct {
+	prefix string
+	tag    string
+}
+
+// NewModeler returns a pointer to a new Modeler, confgiured with the provided map key prefix and
+// struct tag key.
+func NewModeler(prefix string, tag string) *Modeler {
+	return &Modeler{prefix: prefix, tag: tag}
+}
+
+// MapToModel populates the provided model with values from the provided map.
+func (m *Modeler) MapToModel(data map[string]string, out interface{}) error {
+	rv := reflect.ValueOf(out)
+	return m.mapToModel(data, "", rv)
+}
+
+func (m *Modeler) mapToModel(data map[string]string, context string, rv reflect.Value) error {
+	// If rv is invalid (represents a nil litersl), we cannot proceed.
+	if rv.Kind() == reflect.Invalid {
+		return fmt.Errorf("Cannot populate a nil from the map.")
+	}
+	// We also cannot proceed if rv is not a pointer.
+	if rv.Kind() != reflect.Ptr {
+		return fmt.Errorf("Cannot populate non-pointer type %s from the map.", rv.Type())
+	}
+	// Or if it is a pointer to a nil.
+	if rv.IsNil() {
+		return fmt.Errorf("Cannot populate nil %s from the map.", rv.Type())
+	}
+	// At this point, we know we're working with a pointer that points to something.
+	// Get what it points to...
+	elem := rv.Elem()
+	// If the thing it points to isn't a struct, that's also a problem.
+	if elem.Kind() != reflect.Struct {
+		return fmt.Errorf("Cannot populate non-struct-pointer type %s from the map.", rv.Type())
+	}
+	rt := elem.Type()
+	for i := 0; i < rt.NumField(); i++ {
+		rf := rt.Field(i)
+		tagValue := rf.Tag.Get(m.tag)
+		// If no tag value is found...
+		if tagValue == "" {
+			// Just move to the next field.
+			continue
+		}
+		if rf.Type.Kind() == reflect.Ptr || rf.Type.Kind() == reflect.Struct {
```

Comment:
It would fail, but not silently.  All the conditions that need to be met for handing the "top level" struct  pointer need to be satisfied when handling anything we recurse into, else an error is returned.  I can export and better document the errors as you've suggested, but I think that this is the right behavior.  We really can't do anything with anything that isn't a pointer to a struct.